File: src/VisualBasic/CodeCracker/Style/PreferAnyToCountGreaterThanZeroAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Style
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class PreferAnyToCountGreaterThanZeroAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Shared ReadOnly Title As LocalizableString = New LocalizableResourceString(NameOf(Properties.Resources.PreferAnyToCountGreaterThanZeroAnalyzer_Title), Properties.Resources.ResourceManager, GetType(Properties.Resources))
        Friend Shared ReadOnly MessageFormat As LocalizableString = New LocalizableResourceString(NameOf(Properties.Resources.PreferAnyToCountGreaterThanZeroAnalyzer_MessageFormat), Properties.Resources.ResourceManager, GetType(Properties.Resources))
        Friend Const Category = SupportedCategories.Style

        Friend Shared Rule As New DiagnosticDescriptor(
            DiagnosticId.PreferAnyToCountGreaterThanZero.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.PreferAnyToCountGreaterThanZero)
        )

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(Rule)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf AnalyzeNode, SyntaxKind.GreaterThanExpression)
        End Sub

        Private Shared Sub AnalyzeNode(context As SyntaxNodeAnalysisContext)
            If (context.IsGenerated()) Then Return
            Dim binExpression = TryCast(context.Node, BinaryExpressionSyntax)
            If Not binExpression.IsKind(SyntaxKind.GreaterThanExpression) Then Return
            Dim rightSideExpression = TryCast(binExpression.Right, LiteralExpressionSyntax)
            If rightSideExpression Is Nothing Then Return
            If Not rightSideExpression.IsKind(SyntaxKind.NumericLiteralExpression) Then Return
            If rightSideExpression.Token.ToString() <> "0" Then Return
            Dim memberExpression = binExpression.Left.DescendantNodesAndSelf().OfType(Of MemberAccessExpressionSyntax).FirstOrDefault()
            If memberExpression Is Nothing Then Return
            If memberExpression.Name.ToString() <> "Count" Then Return
            Dim memberSymbolInfo = context.SemanticModel.GetSymbolInfo(memberExpression)
            Dim namespaceName = memberSymbolInfo.Symbol.ContainingNamespace.ToString()
            If namespaceName <> "System.Linq" AndAlso namespaceName <> "System.Collections" AndAlso namespaceName <> "System.Collections.Generic" Then Return

            context.ReportDiagnostic(Diagnostic.Create(Rule, binExpression.GetLocation(), GetPredicateString(binExpression.Left)))
        End Sub

        Private Shared Function GetPredicateString(expression As ExpressionSyntax) As String
            Dim predicateString = ""
            If TypeOf expression Is InvocationExpressionSyntax Then
                Dim arguments = TryCast(expression, InvocationExpressionSyntax).ArgumentList
                predicateString = If(arguments?.Arguments Is Nothing, "", If(arguments.Arguments.Count > 0, "predicate", ""))
            End If
            Return predicateString
        End Function
    End Class
End Namespace

File: src/CSharp/CodeCracker/Style/PreferAnyToCountGreaterThanZeroAnalyzer.cs
using CodeCracker.Properties;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Style
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class PreferAnyToCountGreaterThanZeroAnalyzer : DiagnosticAnalyzer
    {
        internal static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.PreferAnyToCountGreaterThanZeroAnalyzer_Title), Resources.ResourceManager, typeof(Resources));
        internal static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.PreferAnyToCountGreaterThanZeroAnalyzer_MessageFormat), Resources.ResourceManager, typeof(Resources));
        internal const string Category = SupportedCategories.Style;

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.PreferAnyToCountGreaterThanZero.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Info,
            isEnabledByDefault: true,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.PreferAnyToCountGreaterThanZero));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.GreaterThanExpression);

        private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var binExpression = context.Node as BinaryExpressionSyntax;
            if (binExpression.IsNotKind(SyntaxKind.GreaterThanExpression)) return;
            var rightSideExpression = binExpression.Right as LiteralExpressionSyntax;
            if (rightSideExpression == null) return;
            if (rightSideExpression.IsNotKind(SyntaxKind.NumericLiteralExpression)) return;
            if (rightSideExpression.Token.ToString() != "0") return;
            if (binExpression.Left == null) return;
            if (binExpression.Left.IsNotKind(SyntaxKind.InvocationExpression, SyntaxKind.SimpleMemberAccessExpression)) return;
            var memberExpression = ((binExpression.Left as InvocationExpressionSyntax)?.Expression ?? binExpression.Left) as MemberAccessExpressionSyntax;
            if (memberExpression == null) return;
            if (memberExpression.Name.ToString() != "Count") return;
            var memberSymbolInfo = context.SemanticModel.GetSymbolInfo(memberExpression);
            var namespaceName = memberSymbolInfo.Symbol?.ContainingNamespace?.ToString();
            if (namespaceName != "System.Linq" && namespaceName != "System.Collections" && namespaceName != "System.Collections.Generic") return;
            context.ReportDiagnostic(Diagnostic.Create(Rule, binExpression.GetLocation(), GetPredicateString(binExpression.Left)));
        }

        private static string GetPredicateString(ExpressionSyntax expression)
        {
            var predicateString = "";
            if (expression.IsKind(SyntaxKind.InvocationExpression))
            {
                var arguments = ((InvocationExpressionSyntax)expression).ArgumentList;
                predicateString = arguments?.Arguments == null ? "" : arguments.Arguments.Count > 0 ? "predicate" : "";
            }
            return predicateString;
        }
    }
}